packages/devtools/src/schema/compiler.ts

```diff
@@ -0,0 +1,328 @@
+      list.minItems = definition.minLength
+    }
+
+    let item: SchemaWithRefs | void
+    switch (definition.item.type) {
+      case 'enum':
+        item = this._compileEnum(objectName, fieldName, definition.item)
+        break
+      case 'object':
+        item = this._compileObjectReference(objectName, fieldName, definition.item)
+        break
+      case 'scalar':
+        item = this._compileScalar(definition.item)
+        break
+    }
+
+    if (item == null) {
+      throw new Error(`Could not compile item schema for list ${fieldName} of object ${objectName}`)
+    }
+    list.items = item.schema
+
+    return { schema: list, refs: item.refs }
+  }
+
+  _compileObjectReference(
+    objectName: string,
+    fieldName: string,
+    definition: ObjectReferenceFieldDefinition
+  ): SchemaWithRefs {
+    if (this.#src.models[definition.name] != null) {
+      throw new Error(
+        `Unsupported reference to model ${definition.name} in field ${fieldName} of object ${objectName}. References can only be made to embedded objects.`
+      )
+    }
+
+    const target = this.#src.objects[definition.name]
+    if (target == null) {
+      throw new Error(
+        `Missing object ${definition.name} referenced in field ${fieldName} of object ${objectName}`
+      )
+    }
+
+    // Ensure object is compiled and injected to definitions record
+    this._compileEmbedObject(definition.name, target)
+
+    return { schema: { $ref: `#/$defs/${definition.name}` }, refs: [definition.name] }
+  }
+
+  _compileEnum(
+    objectName: string,
+    fieldName: string,
+    definition: EnumFieldDefinition
+  ): SchemaWithRefs {
+    const values = this.#src.enums[definition.name]
+    if (values == null) {
+      throw new Error(
+        `Missing enum ${definition.name} referenced in field ${fieldName} of object ${objectName}`
+      )
+    }
+    if (this.#refs[definition.name] == null) {
+      this.#refs[definition.name] = {
+        schema: { type: 'string', title: definition.name, enum: values },
+        refs: [],
+      }
+    }
+    return { schema: { $ref: `#/$defs/${definition.name}` }, refs: [definition.name] }
+  }
+
+  _compileScalar(definition: ScalarFieldDefinition): SchemaWithRefs {
+    const title = definition.schema.title
+    // Scalars without title or that have properties changed from the defaults are injected directly as they are not reusable
+    if (title == null || !isCommonScalar(definition.schema)) {
+      return { schema: definition.schema, refs: [] }
+    }
+
+    // Scalars with title are injected in definitions and referenced
+    if (this.#refs[title] == null) {
+      this.#refs[title] = { schema: definition.schema, refs: [] }
+    }
+    return { schema: { $ref: `#/$defs/${title}` }, refs: [title] }
+  }
+
+  _compileModel(
+    name: string,
+    modelDefinition: ParsedModelDefinition,
+    objectDefinition: ObjectDefinition
+  ): AbstractModelDefinition {
+    if (modelDefinition.action === 'load') {
```

Comment:
Can we have an enum or constants for definition actions?

Reply:
`action` is used as the discriminatory field on the type, it doesn't have other needs so that seems unnecessary.

Reply:
We don't have to export the enum/constants, but I just thought it would still be better to have these strings defined in one place, so that we only need to update them in one place, if we decide to change them :)